tests/integration-tests/tests/storage/snapshots_factory.py

```diff
@@ -0,0 +1,255 @@
+        logging.info("Waiting for the volume to be ready...")
+        while vol.state == u"creating":
+            vol = self.ec2.Volume(vol.id)
+            time.sleep(2)
+        logging.info("Volume ready")
+        return vol
+
+    def _get_security_group_id(self):
+        response = self.boto_client.describe_security_groups(
+            Filters=[{"Name": "group-name", "Values": ["snapshot"]}, {"Name": "vpc-id", "Values": [self.config.vpc_id]}]
+        )
+        if len(response["SecurityGroups"]) == 0:
+            security_group_id = self.boto_client.create_security_group(
+                Description="security group for snapshot instance node", GroupName="snapshot", VpcId=self.config.vpc_id
+            )["GroupId"]
+
+            self.boto_client.authorize_security_group_ingress(
+                GroupId=security_group_id,
+                IpPermissions=[
+                    {"IpProtocol": "tcp", "FromPort": 22, "ToPort": 22, "IpRanges": [{"CidrIp": "0.0.0.0/0"}]}
+                ],
+            )
+        else:
+            security_group_id = response["SecurityGroups"][0]["GroupId"]
+
+        return security_group_id
+
+    def _launch_instance(self, ami_id, subnet):
+        instance = self.ec2.create_instances(
+            ImageId=ami_id,
+            # SubnetId=cluster_config.master_subnet_id,
+            KeyName=self.config.key_name,
+            MinCount=1,
+            MaxCount=1,
+            InstanceType="t2.micro",
+            NetworkInterfaces=[
+                {
+                    "SubnetId": subnet.id,
+                    "DeviceIndex": 0,
+                    "AssociatePublicIpAddress": True,
+                    "Groups": [self._get_security_group_id()],
+                }
+            ],
+            TagSpecifications=[
+                {"ResourceType": "instance", "Tags": [{"Key": "Name", "Value": "pcluster-snapshot-instance"}]}
+            ],
+        )[0]
+        logging.info("Waiting for instance to be running...")
+        while instance.state["Name"] == u"pending":
+            time.sleep(10)
+            instance = self.ec2.Instance(instance.id)
+
+        logging.info("Instance state: %s" % instance.state)
+        logging.info("Public dns: %s" % instance.public_dns_name)
+        return instance
+
+    def _get_amazonlinux_ami(self):
+        # Finds most recent alinux ami in region
+        response = self.boto_client.describe_images(
+            Owners=["amazon"],
+            Filters=[
+                {"Name": "name", "Values": ["amzn-ami-hvm-*"]},
+                {"Name": "description", "Values": ["Amazon Linux AMI*"]},
+                {"Name": "architecture", "Values": ["x86_64"]},
+                {"Name": "root-device-type", "Values": ["ebs"]},
+                {"Name": "state", "Values": ["available"]},
+            ],
+        )
+
+        amis = sorted(response["Images"], key=lambda x: x["CreationDate"], reverse=True)
+        return amis[0]["ImageId"]
+
+    def release_all(self):
```

Comment:
The problem is that we can't release the volume if release_instance fails. That's why I put a @retry do make it do its best. In any case, the instance should be already terminated at that point since this is done before creating the snapshot. I re-added it here just in case. Let's talk if you have a better idea.